mypy/messages.py

```diff
@@ -998,6 +999,184 @@ def untyped_decorated_function(self, typ: Type, context: Context) -> None:
+                      'conflicts:'.format(self.format(subtype)), context)
+            for name, got, exp in conflict_types[:MAX_ITEMS]:
+                if (not isinstance(exp, (CallableType, Overloaded)) or
+                        not isinstance(exp, (CallableType, Overloaded))):
+                    self.note('{}: expected {}, got {}'.format(name,
+                                                               *self.format_distinctly(exp, got)),
+                              context, offset=OFFSET)
+                else:
+                    self.note('Expected:', context, offset=OFFSET)
+                    if isinstance(exp, CallableType):
+                        self.note(self.pretty_callable(exp), context, offset=2 * OFFSET)
+                    else:
+                        assert isinstance(exp, Overloaded)
+                        pretty_overload(exp)
+                    self.note('Got:', context, offset=OFFSET)
+                    if isinstance(got, CallableType):
+                        self.note(self.pretty_callable(got), context, offset=2 * OFFSET)
+                    else:
+                        assert isinstance(got, Overloaded)
+                        pretty_overload(got)
+            print_more(conflict_types)
+
+        # Report flag conflicts (i.e. settable vs read-only etc.)
+        conflict_flags = get_bad_flags(subtype, supertype)
+        for name, subflags, superflags in conflict_flags[:MAX_ITEMS]:
+            if IS_CLASSVAR in subflags and IS_CLASSVAR not in superflags:
+                self.note('Protocol member {}.{} expected instance variable,'
+                          ' got class variable'.format(supertype.type.name(), name), context)
+            if IS_CLASSVAR in superflags and IS_CLASSVAR not in subflags:
+                self.note('Protocol member {}.{} expected class variable,'
+                          ' got instance variable'.format(supertype.type.name(), name), context)
+            if IS_SETTABLE in superflags and IS_SETTABLE not in subflags:
+                self.note('Protocol member {}.{} expected settable variable,'
+                          ' got read-only attribute'.format(supertype.type.name(), name), context)
+            if IS_CLASS_OR_STATIC in superflags and IS_CLASS_OR_STATIC not in subflags:
+                self.note('Protocol member {}.{} expected class or static method'
+                          .format(supertype.type.name(), name), context)
+        print_more(conflict_flags)
+
+    def pretty_callable(self, tp: CallableType) -> str:
+        """Return a nice easily-readable representation of a callable type.
+        For example:
+            def [T <: int] f(self, x: int, y: T) -> None
+        """
+        s = ''
+        bare_asterisk = False
+        for i in range(len(tp.arg_types)):
+            if s:
+                s += ', '
+            if tp.arg_kinds[i] in (ARG_NAMED, ARG_NAMED_OPT) and not bare_asterisk:
```

Comment:
Consider adding test cases for pretty printing various kinds of arguments -- the current tests didn't look exhaustive.